### ISSUE TITLE
Rework of the bot--and some other info

### ISSUE THREAD

**Issue by spergmoment** [spergmoment/FederalBot#33]
I have a LOT to cover here, so let's get straight into it.

I am currently planning an absolutely huge rework of the bot. This will pretty much change the whole fundamentals of the bot. Don't worry--it'll do everything the same way, just much better, with some added features on as well!

You may have noticed a drop in commits recently, and that's because I've been testing out some Sequelize stuff to make it PERFECT.

This rework will pretty much fix every single problem the bot has and more. I will be adding so many features to it. So,  let's see whats changed!

# EVENTS
This is one of the smaller parts but is still big nonetheless. SO, here goes:
* Fix up member remove event
* Add a few events I didn't already have
* Rework many of the events to properly detect more info (i.e. updates to certain aspects of the server)
* Polish the message delete event

# MONEY SYSTEM
Despite how big it sounds, this is only a small part of the update as well.
* Implement a "cash" system with things such as gambling, items, and more!
## LAWYER SYSTEM
Part of the money system!
* A more effective way to handle court cases.
* Easy way for defendants to get the help they need!
* Pleas, automatic punishments and more, it's like an entirely new court system! (as if I haven't had to do that a thousand times before...)

# COMMAND FIXES
Some general command fixes. Nothing specific here.

# PROMISES
Look at any of the commands. Notice all the .then() callbacks.
* Those .then() callbacks will be gone!
* Cleaner, modernized code, without all the messy indents!
* Even better performance! No more taking 2 years to respond to something!

# v12 CONFORMING
Oh boy, this is the BIGGEST part of the update yet.
* Check issue #29 for more info.

# GENERAL MODERNIZATION
The code will just be generally "modernized" to better fit ES8. This goes with Promises, v12 conforming, and the money system.

# REFORMATTING
I will go through every nook and cranny and make sure EVERYTHING is formatted PERFECTLY!
* Better overall code readability
* Easy-to-read logs
* Better overall logging features

This is by far my most ambitious project yet. Development on this gargantuan update starts in 2 days. Until then, I will be preparing, both myself and the bot. I will also look for a better host.

Thank you for reading!

TL;DR: I will generally be reformatting, modernizing, and fixing almost the entirety of the bot, and add some new features too!

**Comment by spergmoment** [comment]
finished promise callback stuff

**Comment by spergmoment** [comment]
UPDATE 1:
Suprisingly, v12 conforming went pretty well! All works great. Anyway, changed up a few things so here goes:
- Conformed to v12 fully
- Formatted logs a bit
- Formatted the code
- Reworked the help command (again) so that even the menus are dynamic
- Put in ES10 conformation implementations
- Fixed some commands
- Reworked resign to be dynamic
- Updated bill to better fit the bot's style
- Removed the module command
- Added botinfo and guild info commands
- Added a DM-only ticket feature
- Added a response on a non-ticket/help DM
# BREAKING CHANGES
- Invites have some issues
- Nominate is buggy again. I'm remaking the command anyway so yeah
- Ping has some issues, due to the client calling the websocket, which makes an extra call to the api, ratelimiting the bot and skewing results
# FIXES
- guildMemberRemove no longer errors out the process
- guildMemberAdd formatting fixed
- Fully conformed to string lits (Except for eval--more info soon)
- Fixed 45 process errors
- Fixed 12 host errors
- Changed DNS for better connection
- Ping no longer crashes the process on a failed ping attempt

This is actually really fun to do. I don't know why.

Anyway, I'll be pushing aside events for a while.

**Comment by spergmoment** [comment]
Clarification on eval:

I will also be reworking this command, and for it to work the way I need, I *can not* use string literals. I simply will have to deal with them for now. Luckily, since eval is part of the base API and not in any way related to Discord, I am able to keep it like this.

**Comment by spergmoment** [comment]
# UPDATE 2:
This one is a bit smaller. However, it's still relatively big. So, here goes:
# NEW FEATURES
- Nominate is dynamic
- Impeach is dynamic
- Logs are more dynamic
- Eval is now asynced, meaning awaits can be used without declaring an async function
- Added a few aliases
# BREAKING CHANGES
- Nominate statically calls methods sometimes
- Impeach statically calls methods sometimes
- Menu 2 of help doesn't show up depending on the current performance and memory usage
# FIXES
- Invites work (for the most part)
- Ping no longer takes an extra API call
- Clear no longer clears a sporadic amount of messages
- Ratelimit errors have been significantly reduced
- Fixed 92 process errors
- Fixed 3 host errors
# KNOWN ISSUES
These were known, but never documented because they are too small.
- Logs often receive the wrong audit logs
- Help may very rarely receive the wrong info
- Ticket can send a message to the botowners that is over 2000 characters, erroring -- use an embed

I'll be ironing out many of the issues you see here.

Additionally, I'll be looking into loggerv3 and getting some ideas from there for events. Soon

**Comment by spergmoment** [comment]
# UPDATE 3
No real changes here, just some fixes.
# FIXED
- Ticket uses an embed now
- Help no longer receives incorrect info (fixed dynamics)
- Nominate and Impeach are less likely to call static methods, as well as Resign
- Fixed second menu of help
- Polished up invites
- 4 process errors, 3 host errors, and 23 connection lookup errors fixed
- Fixed SQLite3 processing, install, and build errors

Next update will be better, I promise.

**Comment by spergmoment** [comment]
Next update is command reworks, which has already proved to be incredibly hard to do.

**Comment by spergmoment** [comment]
I'll be postponing this for a little bit. I currently have a ton of other projects to do.

Estimated time of return is in 1-3 days.

**Comment by spergmoment** [comment]
Also, earlier, I implemented a fix for wrong audit logs. It simply doesn't log it if it happened a while ago (i.e. it didn't properly fetch the audit log), which as of now I've set to 30 seconds. This is subject to change.

**Comment by spergmoment** [comment]
ive added a (very experimental) channel update thing

ill be pushing the actual code in a little while

**Comment by spergmoment** [comment]
federalbot is done ig

**Comment by spergmoment** [comment]
nvm
